config_system/config_system/general.py

```diff
@@ -43,105 +43,176 @@ def check_depends(depends, value):
-                    check_depends(depends[2], value))
-        else:
-            return False
-    return depends == value
+    assert type(depends) == tuple
+    assert len(depends) in [2, 3]
+
+    if depends[0] == 'and':
+        return (check_depends(depends[1], value) or
+                check_depends(depends[2], value))
+    elif depends[0] == 'or':
+        return (check_depends(depends[1], value) and
+                check_depends(depends[2], value))
+    elif depends[0] == 'word':
+        return depends[1] == value
+    return False
 
 
-def value_of(depends):
-    if depends is None:
-        return "y"
-    if type(depends) == tuple:
-        if len(depends) == 3:
-            left = value_of(depends[1])
-            right = value_of(depends[2])
-            if type(left) != type(right):
-                return 'n'
-            elif depends[0] == 'and':
-                if left == 'y' and right == 'y':
-                    return 'y'
-                return 'n'
-            elif depends[0] == 'or':
-                if left == 'y' or right == 'y':
-                    return 'y'
-                return 'n'
-            elif depends[0] == '=':
-                if left == right:
-                    return 'y'
-                return 'n'
-            elif depends[0] == '!=':
-                if left != right:
-                    return 'y'
-                return 'n'
-            elif depends[0] == '>':
-                if left > right:
-                    return 'y'
-                return 'n'
-            elif depends[0] == '>=':
-                if left >= right:
-                    return 'y'
-                return 'n'
-            elif depends[0] == '<':
-                if left < right:
-                    return 'y'
-                return 'n'
-            elif depends[0] == '<=':
-                if left <= right:
-                    return 'y'
-                return 'n'
-        elif depends[0] == 'not':
-            if value_of(depends[1]) == 'y':
-                return 'n'
-            return 'y'
-        elif depends[0] == 'string':
-            return depends[1]
-        elif depends[0] == 'number':
-            return depends[1]
-        else:
-            raise Exception("Unexpected depend list: " + str(depends))
-    return configuration['config'][depends]['value']
+# Operators where the result is not boolean
+ARITH_SET = {"+", "-"}
+
+
+def _expr_value(e):
+    """Evaluate the value of the input expression.
+    This isn't expected to use conditional operators.
+    """
+    assert type(e) == tuple
+    assert len(e) in [2, 3]
+    if len(e) == 3:
+        left = _expr_value(e[1])
+        right = _expr_value(e[2])
+        if type(left) != type(right):
+            raise TypeError("'{}' operator is not valid with mixed types".format(e[0]))
+        elif left in ['y', 'n'] or right in ['y', 'n']:
+            raise TypeError("'{}' operator is not valid on booleans".format(e[0]))
```

Comment:
Won't this incorrectly error if we wanted to append _strings_ where one of the operands is 'y' or 'n'?

I guess we could leave this commented out until we can convert booleans to using actual `bool` types.

Or actually, could `_expr_value()` return a tuple of `(type, value)` instead? Then we discard the type in the wrapper `expr_value`?

Reply:
Yes, this will behave incorrectly if you want a string to have the value 'y' or 'n'. I don't believe we have instances of this in user configs, so we should be OK for now. We should just go straight to handling the booleans as a distinct type. We can do this in a follow up commit, as it will affect a fair amount of unrelated code.

Reply:
OK - I've seen a nearly-ready version of a patch to go directly to actual bools, so this is bearable for now.